src/frontend/Ast_to_Mir.ml

```diff
@@ -160,13 +171,16 @@ let mkfor bodyfn iteratee smeta =
     type. In this way we recursively create for loops that loop over
     the outermost layers first.
 *)
-let rec for_scalar bodyfn var smeta =
-  match var.emeta.mtype with
-  | UInt | UReal -> bodyfn var
-  | UVector | URowVector | UMatrix -> mkfor bodyfn var smeta
-  | UArray _ -> mkfor (fun e -> for_scalar bodyfn e smeta) var smeta
-  | UFun _ | UMathLibraryFunction ->
-      raise_s [%message "Can't for over " (var : mtype_loc_ad with_expr)]
+let rec for_scalar st bodyfn var smeta =
+  match st with
+  | SInt | SReal -> bodyfn var
+  | SVector d | SRowVector d -> mkfor d bodyfn var smeta
+  | SMatrix (d1, d2) ->
+      mkfor
+        { expr= FunApp (StanLib, string_of_operator Times, [d1; d2])
```

Comment:
You have a single loop variable here, so I guess it's the responsibility of bodyfn to index correctly into the matrix using the single index?